libsolidity/analysis/ControlFlowAnalyzer.cpp

```diff
@@ -134,36 +178,48 @@ void ControlFlowAnalyzer::checkUninitializedAccess(CFGNode const* _entry, CFGNod
 
 		for (auto const* variableOccurrence: uninitializedAccessesOrdered)
 		{
+			VariableDeclaration const& varDecl = variableOccurrence->declaration();
+
 			SecondarySourceLocation ssl;
 			if (variableOccurrence->occurrence())
-				ssl.append("The variable was declared here.", variableOccurrence->declaration().location());
+				ssl.append("The variable was declared here.", varDecl.location());
 
-			bool isStorage = variableOccurrence->declaration().type()->dataStoredIn(DataLocation::Storage);
-			bool isCalldata = variableOccurrence->declaration().type()->dataStoredIn(DataLocation::CallData);
+			bool isStorage = varDecl.type()->dataStoredIn(DataLocation::Storage);
+			bool isCalldata = varDecl.type()->dataStoredIn(DataLocation::CallData);
 			if (isStorage || isCalldata)
 				m_errorReporter.typeError(
 					3464_error,
 					variableOccurrence->occurrence() ?
 						*variableOccurrence->occurrence() :
-						variableOccurrence->declaration().location(),
+						varDecl.location(),
 					ssl,
 					"This variable is of " +
 					string(isStorage ? "storage" : "calldata") +
 					" pointer type and can be " +
 					(variableOccurrence->kind() == VariableOccurrence::Kind::Return ? "returned" : "accessed") +
 					" without prior assignment, which would lead to undefined behaviour."
 				);
-			else if (!_emptyBody && variableOccurrence->declaration().name().empty())
+			else if (!_emptyBody && varDecl.name().empty())
+			{
+				if (!m_previousVariableWarnings.emplace(&varDecl).second)
+					continue;
+
 				m_errorReporter.warning(
 					6321_error,
-					variableOccurrence->declaration().location(),
-					"Unnamed return variable can remain unassigned. Add an explicit return with value to all non-reverting code paths or name the variable."
+					varDecl.location(),
+					(
+						!_contract || _contract == _function.annotation().contract ?
+						"U" :
+						"When called using contract \"" + _contract->name() + "\" the u"
+					) +
+					"nnamed return variable can remain unassigned. Add an explicit return with value to all non-reverting code paths or name the variable."
```

Comment:
How about this? Currently it sounds as if the variable is being called, not the function. Also the reversed condition is a bit easier to parse for me:
```suggestion
					"Unnamed return variable can remain unassigned " +
					(
						_contract && _contract != _function.annotation().contract ?
						"when the function is called within contract \"" + _contract->name() + ". ":
						". "
					) +					
					"Add an explicit return with value to all non-reverting code paths or name the variable."
```

Reply:
lgtm